### PR TITLE
Add OpenID-Connect support 

### PR DESCRIPTION
**Pull request by jvlcek**
https://www.pivotaltracker.com/n/projects/1610127/stories/121849185

The changes introduced by this PR will provide support for enabling OpenID-Connect
authentication.

### There are four fundamental changes:

1. Add the oidc.rb to support OpenID-Connect config map generation.
2. Update the config_map.rb to accommodate for an optional additional data name value hash.
3. Move the host option from base.rb to the individual provider type implementations as needed, because oidc does not need host.
4. Add specs to ensure the required_options are returned correctly after change #3 

### This PR also requires PRs:
https://github.com/ManageIQ/manageiq-ui-classic/pull/2855
https://github.com/ManageIQ/manageiq-pods/pull/251 and back ported to the hammer branch
https://github.com/ManageIQ/container-httpd/pull/33

### CODE REVIEW

**Comment by jvlcek** [comment]
@miq-bot add_label enhancement

**Comment by jvlcek** [comment]
@miq-bot assign @abellotti

**Comment by carbonin** [comment]
@abellotti can you take a look at this when you get a chance? I'm not sure anyone else has merge rights here.

File: README-oidc.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+## Usage for the `oidc` auth-type:
+
+```
+$ httpd_configmap_generator oidc --help
+Options:
+  -o, --output=<s>                Configuration map file to create
+  -i, --oidc-url=<s>              OpenID-Connect Provider URL
+  -c, --oidc-client-id=<s>        OpenID-Connect Provider Client ID
+  -l, --oidc-client-secret=<s>    OpenID-Connect Provider Client Secret
```
Comment (abellotti) [inline README-oidc.md]: can these short options be -u for url, -i for id and -s for secret ?

File: lib/httpd_configmap_generator/base/config_map.rb [hunk 2]
```diff
@@ -16,11 +16,11 @@ def initialize(opts = {})
       @config_map = template
     end
 
-    def generate(auth_type, realm, file_list)
+    def generate(auth_type, realm = "undefined", file_list = nil, additional_data = {})
```
Comment (abellotti) [inline lib/httpd_configmap_generator/base/config_map.rb]: maybe rename additional_data as just metadata ?

File: lib/httpd_configmap_generator/base/config_map.rb [hunk 3]
```diff
@@ -135,7 +135,7 @@ def file_entry_spec(source_file, target_file = nil, mode = nil)
       }
     end
 
-    def update_configuration(file_specs)
+    def update_configuration(file_specs, extra_data={})
```
Comment (abellotti) [inline lib/httpd_configmap_generator/base/config_map.rb]: metadata={}

File: lib/httpd_configmap_generator/base/config_map.rb [hunk 4]
```diff
@@ -157,9 +157,9 @@ def search_file_entry(target_file)
       entry ? entry.first.split('=')[1].strip.split(' ') : nil
     end
 
-    def define_configuration(file_specs)
+    def define_configuration(file_specs, extra_data={})
```
Comment (abellotti) [inline lib/httpd_configmap_generator/base/config_map.rb]: same

File: README-oidc.md (hunk 1 above)
Comment (jvlcek) [reply README-oidc.md]: @abellotti Good catch! I thought I had fixed those. Will do!

File: lib/httpd_configmap_generator/base/config_map.rb (hunk 2 above)
Comment (jvlcek) [reply lib/httpd_configmap_generator/base/config_map.rb]: @abellotti 👍 Will do.

**Comment by abellotti** [comment]
This is looking great.  Thanks @jvlcek for pushing this to the finish line. LGTM!!  👍

**Pull request merged**